JsonView with ExceptionHandler throws 500 error in 4.1.2

### Expected Behavior

I call a rest method with the @JsonView annotation. The method throws an error. The error gets handled by my error handler and returns the appropriate response status and body.

### Actual Behaviour

Instead of my handled error response being returned I get a 500 error instead. In the logs I see an exception is thrown saying that the the JsonError is incompatible with the expected return type.

```
09:07:08.770 [io-executor-thread-3] ERROR i.m.http.server.RouteExecutor - Unexpected error occurred: Error encoding object [Not Found] to JSON: Incompatible types: declared root type ([simple type, class test_email_validation.MyObj]) vs `io.micronaut.http.hateoas.JsonError`
io.micronaut.http.codec.CodecException: Error encoding object [Not Found] to JSON: Incompatible types: declared root type ([simple type, class test_email_validation.MyObj]) vs `io.micronaut.http.hateoas.JsonError`
	at io.micronaut.json.codec.MapperMediaTypeCodec.encode(MapperMediaTypeCodec.java:268)
	at io.micronaut.http.server.netty.jackson.JsonViewServerFilter.lambda$doFilter$1(JsonViewServerFilter.java:99)
	at reactor.core.publisher.MonoCallable.call(MonoCallable.java:72)
	at reactor.core.publisher.FluxSubscribeOnCallable$CallableSubscribeOnSubscription.run(FluxSubscribeOnCallable.java:227)
	at io.micronaut.core.propagation.PropagatedContext.lambda$wrap$3(PropagatedContext.java:211)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28)
	at io.micronaut.core.propagation.PropagatedContext.lambda$wrap$4(PropagatedContext.java:228)
	at io.micronaut.core.propagation.PropagatedContext.lambda$wrap$4(PropagatedContext.java:228)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Incompatible types: declared root type ([simple type, class test_email_validation.MyObj]) vs `io.micronaut.http.hateoas.JsonError`
	at com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:77)
	at com.fasterxml.jackson.databind.SerializerProvider.reportBadDefinition(SerializerProvider.java:1308)
	at com.fasterxml.jackson.databind.SerializerProvider._reportIncompatibleRootType(SerializerProvider.java:1399)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:380)
	at com.fasterxml.jackson.databind.ObjectWriter$Prefetch.serialize(ObjectWriter.java:1568)
	at com.fasterxml.jackson.databind.ObjectWriter._writeValueAndClose(ObjectWriter.java:1273)
	at com.fasterxml.jackson.databind.ObjectWriter.writeValueAsBytes(ObjectWriter.java:1163)
	at io.micronaut.jackson.databind.JacksonDatabindMapper.writeValueAsBytes(JacksonDatabindMapper.java:209)
	at io.micronaut.json.codec.MapperMediaTypeCodec.encode(MapperMediaTypeCodec.java:265)
	... 12 common frames omitted
```

### Steps To Reproduce

1.  Create a GET rest endpoint
2. Annotate rest endpoint with JSON view and the appropriate view class
3. rest endpoint should declare a return type other than void.
4. Throw a custom exception
5. Create a ExceptionHandler for the custom exception
6. ExceptionHandler should generate a response with a body of type JsonError.
7. Startup the application, hit the endpoint that throws the custom exception. This should result in a 500 error because JsonError is not of the same type as the rest endpoint return type.

### Environment Information

- Mac
- Java 17
- Maven 3.9.4
- Micronaut 4.1.2

### Example Application

https://github.com/kevind-wgu/micronaut-jsonview-bug

### Version

4.1.2

--- a/http-server-netty/src/main/java/io/micronaut/http/server/netty/jackson/JsonViewServerFilter.java
+++ b/http-server-netty/src/main/java/io/micronaut/http/server/netty/jackson/JsonViewServerFilter.java
@@ -86,7 +86,7 @@ public class JsonViewServerFilter implements Ordered {
                 .flatMap(AnnotationValue::classValue);
             if (viewClass.isPresent() &&
                 // if this is an error response, the response body does not come from the controller
-                response.getAttribute(HttpAttributes.EXCEPTION).isEmpty()) {
+                !response.getAttributes().contains(HttpAttributes.EXCEPTION)) {
 
                 final Optional<?> optionalBody = response.getBody();
                 if (optionalBody.isPresent()) {

--- a/http-server-netty/src/main/java/io/micronaut/http/server/netty/jackson/JsonViewServerFilter.java
+++ b/http-server-netty/src/main/java/io/micronaut/http/server/netty/jackson/JsonViewServerFilter.java
@@ -84,7 +84,10 @@ public class JsonViewServerFilter implements Ordered {
         if (routeInfo != null) {
             final Optional<Class<?>> viewClass = routeInfo.findAnnotation(JsonView.class)
                 .flatMap(AnnotationValue::classValue);
-            if (viewClass.isPresent()) {
+            if (viewClass.isPresent() &&
+                // if this is an error response, the response body does not come from the controller
+                response.getAttribute(HttpAttributes.EXCEPTION).isEmpty()) {
+
                 final Optional<?> optionalBody = response.getBody();
                 if (optionalBody.isPresent()) {
                     Object body = optionalBody.get();

--- a/http-server-netty/src/main/java/io/micronaut/http/server/netty/jackson/JsonViewServerFilter.java
+++ b/http-server-netty/src/main/java/io/micronaut/http/server/netty/jackson/JsonViewServerFilter.java
@@ -86,7 +86,7 @@ public class JsonViewServerFilter implements Ordered {
                 .flatMap(AnnotationValue::classValue);
             if (viewClass.isPresent() &&
                 // if this is an error response, the response body does not come from the controller
-                !response.getAttributes().contains(HttpAttributes.EXCEPTION)) {
+                !response.getAttributes().contains(HttpAttributes.EXCEPTION.toString())) {
 
                 final Optional<?> optionalBody = response.getBody();
                 if (optionalBody.isPresent()) {

--- a/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/jackson/JsonViewController.groovy
+++ b/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/jackson/JsonViewController.groovy
@@ -87,7 +87,7 @@ class JsonViewController {
         throw new Failure1()
     }
 
-    //@JsonView(Views.Public)
+    @JsonView(Views.Public)
     @Get("/failing-with-route")
     TestModel failingWithRoute() {
         throw new Failure2()

--- a/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/jackson/JsonViewController.groovy
+++ b/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/jackson/JsonViewController.groovy
@@ -17,10 +17,13 @@ package io.micronaut.http.server.netty.jackson
 
 import com.fasterxml.jackson.annotation.JsonView
 import io.micronaut.http.HttpResponse
+import io.micronaut.http.HttpStatus
 import io.micronaut.http.annotation.Body
 import io.micronaut.http.annotation.Controller
+import io.micronaut.http.annotation.Error
 import io.micronaut.http.annotation.Get
 import io.micronaut.http.annotation.Post
+import io.micronaut.http.hateoas.JsonError
 import reactor.core.publisher.Flux
 import reactor.core.publisher.Mono
 
@@ -77,4 +80,24 @@ class JsonViewController {
         assert model.password == null
         return HttpResponse.ok(model)
     }
+
+    @JsonView(Views.Public)
+    @Get("/failing")
+    TestModel failing() {
+        throw new Failure1()
+    }
+
+    //@JsonView(Views.Public)
+    @Get("/failing-with-route")
+    TestModel failingWithRoute() {
+        throw new Failure2()
+    }
+
+    @Error
+    HttpResponse<JsonError> errorRoute(Failure2 f) {
+        return HttpResponse.status(HttpStatus.INTERNAL_SERVER_ERROR).body(new JsonError("failure2"));
+    }
+
+    static class Failure1 extends Exception {}
+    static class Failure2 extends Exception {}
 }

--- a/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/jackson/JsonViewServerFilterSpec.groovy
+++ b/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/jackson/JsonViewServerFilterSpec.groovy
@@ -138,10 +138,10 @@ class JsonViewServerFilterSpec extends Specification {
         f.response.getBody(JsonError).get().message == "Internal Server Error"
 
         when:
-        client.toBlocking().exchange(HttpRequest.GET('/jsonview/failing-with-route'), TestModel)
+        client.toBlocking().exchange(HttpRequest.GET('/jsonview/failing-with-route'))
 
         then:
-        def e = thrown HttpClientResponseException
+        HttpClientResponseException e = thrown()
         e.response.getBody(JsonError).get().message == "failure2"
     }
 }

--- a/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/jackson/JsonViewServerFilterSpec.groovy
+++ b/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/jackson/JsonViewServerFilterSpec.groovy
@@ -20,6 +20,8 @@ import io.micronaut.core.type.Argument
 import io.micronaut.http.HttpRequest
 import io.micronaut.http.HttpResponse
 import io.micronaut.http.client.HttpClient
+import io.micronaut.http.client.exceptions.HttpClientResponseException
+import io.micronaut.http.hateoas.JsonError
 import io.micronaut.runtime.server.EmbeddedServer
 import spock.lang.AutoCleanup
 import spock.lang.Shared
@@ -125,5 +127,21 @@ class JsonViewServerFilterSpec extends Specification {
         rsp.body().lastName == JsonViewController.TEST_MODEL.lastName
         rsp.body().birthdate == null
         rsp.body().password == null
+    }
+
+    def "errors are written properly"() {
+        when:
+        client.toBlocking().exchange(HttpRequest.GET('/jsonview/failing'), TestModel)
+
+        then:
+        def f = thrown HttpClientResponseException
+        f.response.getBody(JsonError).get().message == "Internal Server Error"
+
+        when:
+        client.toBlocking().exchange(HttpRequest.GET('/jsonview/failing-with-route'), TestModel)
+
+        then:
+        def e = thrown HttpClientResponseException
+        e.response.getBody(JsonError).get().message == "failure2"
     }
 }

